_posts/2025-04-11-Hybrid-Search-Options-And-Practice.md

```diff
@@ -0,0 +1,275 @@
+* `search_pipeline`: Applies the RRF score combination
+
+
+How RRF works:
+
+
+1. **Sort documents by score**: Each query method sorts documents by score on every shard.
+2. **Assign rank positions**: Documents are ranked based on score for each query.
+3. **Apply the RRF formula**: The RRF score is computed using the following formula:
+
+```
+rankScore(document_i) = sum((1/(k + query_1_rank), (1/(k + query_2_rank), ..., (1/(k + query_j_rank)))
+
+```
+
+In this formula, `k` is a rank constant, and `query_j_rank` represents the ranking of a document in a particular query method. The example in the following diagram applies this formula using the default rank constant of 60.
+
+1. **Add rank contributions**: Rank calculations are combined, and documents are sorted by decreasing rank score.
+2. **Return the top results**: The highest-ranked documents are retrieved based on the query size.
+
+**When to use Rank Fusion:**
+
+* **Heterogeneous Score Distributions:** Rank fusion (especially RRF) excels when BM25 and semantic models produce scores on incompatible scales or with outliers. It avoids complex score calibration by using result ranks only. This yields *stable rankings* even if one query method’s scores would otherwise dominate after normalization.
+* **No Tuning or Calibration Needed:** RRF is a *“plug & play”* method – it requires no pre-training, weight tuning, or knowledge of score ranges. In general-purpose search systems (with diverse queries and content), it’s often impractical to hand-tune weighting for every scenario. Rank fusion performs robustly out-of-the-box, making it a strong default when labeled data for calibration is unavailable
+* **Robust to Outliers and Domain Shifts:** Because RRF ignores absolute score magnitude, an extreme scoring outlier won’t skew the results. This is useful in general collections where some queries or documents might produce anomalously high scores under one model. Similarly, if the data distribution or query mix changes over time, RRF remains stable without needing re-calibration, since it always works off relative rank positions
+
+### Conclusion
+
+The evolution of search techniques in OpenSearch demonstrates the platform's versatility and power in addressing diverse search requirements. While traditional lexical search methods like BM25 continue to provide fast and precise keyword matching, and semantic search offers deep contextual understanding, it is the hybrid search approach that truly revolutionizes information retrieval. Hybrid search in OpenSearch represents a powerful synthesis of these techniques, addressing the limitations of each individual method. 
+
+#### Considerations for Implementation
+
+* **Data Characteristics**: The nature of your data and typical user queries should guide your choice of search technique.
+* **Performance Requirements**: Consider the trade-offs between search accuracy and computational resources, especially for large-scale applications.
```

Comment:
what exactly do you mean here? Is this about hybrid query vs doing hybrid search outside OS, or it's one normalization/combination technique vs another?

Reply:
trade-offs between different searches like bm25, semantic and hybrid based on resource computation